Updated installation, updation, migration docs (#244)
rancher/longhorn-product-docs: docs/version-1.9/modules/en/pages/migration/migration.adoc

=== Remove `longhorn-crd` Helm Release while Keeping CRDs

To remove the `longhorn-crd` Helm release from your system while ensuring the Longhorn CRDs themselves are preserved, follow these steps:

. Check the current version of your `longhorn-crd` Helm release:
+
[,shell]
----
helm list --all-namespaces | grep longhorn-crd
----
+
The output will show your `longhorn-crd` release details, for example:
+
[,bash]
----
longhorn-crd                                     	longhorn-system    	1       	2025-07-21 02:58:38.192579711 +0000 UTC	deployed	longhorn-crd-106.2.1+up1.8.2                                                                           	v1.8.2
----
+
In this example, the release version is `106.2.1+up1.8.2`.

. Clone the `rancher/charts` repository containing the specific version of the `longhorn-crd` chart that matches your current release (for example, `106.2.1+up1.8.2` for Longhorn v1.8.2):
+
[,shell]
----
`+git clone https://github.com/rancher/charts.git+`
----
+
Navigate into the directory corresponding to your `longhorn-crd` chart version within the cloned repository.
+

> Or, you can execute this link:{attachmentsdir}/v1.9.1-longhorn-crd.sh[script] to `longhorn-crd` chart.

Reviewer: Rename the script `v1.9.1-longhorn-crd.sh` to `download-longhorn-crd-chart.sh`

[CustomResourceDefinition] backingimagemanagers.longhorn.io
[CustomResourceDefinition] nodes.longhorn.io
[CustomResourceDefinition] orphans.longhorn.io
[CustomResourceDefinition] recurringjobs.longhorn.io
[CustomResourceDefinition] replicas.longhorn.io
[CustomResourceDefinition] settings.longhorn.io
[CustomResourceDefinition] sharemanagers.longhorn.io
[CustomResourceDefinition] snapshots.longhorn.io
[CustomResourceDefinition] supportbundles.longhorn.io
[CustomResourceDefinition] systembackups.longhorn.io
[CustomResourceDefinition] systemrestores.longhorn.io
[CustomResourceDefinition] backingimages.longhorn.io
[CustomResourceDefinition] volumeattachments.longhorn.io
[CustomResourceDefinition] volumes.longhorn.io
[CustomResourceDefinition] backupbackingimages.longhorn.io
[CustomResourceDefinition] backups.longhorn.io
[CustomResourceDefinition] backuptargets.longhorn.io
[CustomResourceDefinition] backupvolumes.longhorn.io
[CustomResourceDefinition] engineimages.longhorn.io
[CustomResourceDefinition] engines.longhorn.io
[CustomResourceDefinition] instancemanagers.longhorn.io
----

=== Replace `longhorn-crd` with `longhorn` in Longhorn CRDs

> After ensuring the `longhorn-crd` Helm release is uninstalled but the CRDs are kept, you need to update the ownership labels on the existing Longhorn CRDs to prepare for the main `longhorn` Helm chart. Apply the following link:{attachmentsdir}/v1.9.1-longhorn-main.sh[script] to perform the replacement.

Reviewer: Rename `v1.9.1-longhorn-main.sh` to `migrate-crd-ownership.sh`

= Migrating to {longhorn-product-name}
	
[WARNING]
====
It is recommended to create a xref:snapshots-backups/system-backups/create-system-backup.adoc[Longhorn system backup] before performing the upgrade. This ensures that all critical resources, such as volumes and backing images, are backed up and can be restored in case any issues arise.
====

> 

Reviewer 2: @sushant-suse Please add `[NOTE]` like what you did in docs/version-1.9/modules/en/pages/installation-setup/installation/install-using-helm.adoc. Need to include `=== tools` and `=== Authentication`.

=== Authentication

* To authenticate Helm with the SUSE Application Collection OCI chart repository, refer to the https://docs.apps.rancher.io/get-started/authentication/#helm[Helm authentication documentation].
* To allow Kubernetes to pull images from the registry, follow the https://docs.apps.rancher.io/get-started/authentication/#kubernetes[Kubernetes authentication documentation] to create a secret and set the `global.imagePullSecrets` chart value.

== Migrating Longhorn Deployed via Helm To {longhorn-product-name}

If your existing Longhorn installation was deployed via Helm, you can verify it by running `helm list --all-namespaces`. By default, the Helm release used to install Longhorn is named `longhorn`, but it may also be a custom name defined by the user.
	
Next, optionally create a `values.yaml` file with any customized settings you want to apply during the upgrade.

Proceed with the upgrade to {longhorn-product-name} using the {longhorn-product-name} chart:

[,shell]
----
helm upgrade longhorn oci://dp.apps.rancher.io/charts/suse-storage \
	--namespace longhorn-system \
	--version <new-version> \

> 	--set global.imagePullSecrets=<PULL_IMAGE_SECRET> \

Reviewer 2: ```suggestion
	--set global.imagePullSecrets=<PULL_IMAGE_SECRET>
```

----

To migrate the existing Rancher Longhorn installation to {longhorn-product-name}, you need to delete the `longhorn-crd` Helm release without removing the CRDs, and then upgrade the `longhorn` Helm release by following the steps below. 

=== Remove `longhorn-crd` Helm Release while Keeping CRDs

To remove the `longhorn-crd` Helm release from your system while ensuring the Longhorn CRDs themselves are preserved, follow these steps:

. Check the current version of your `longhorn-crd` Helm release:
+
[,shell]
----
helm list --all-namespaces | grep longhorn-crd
----
+
The output will show your `longhorn-crd` release details, for example:
+
[,bash]
----
longhorn-crd                                     	longhorn-system    	1       	2025-07-21 02:58:38.192579711 +0000 UTC	deployed	longhorn-crd-106.2.1+up1.8.2                                                                           	v1.8.2
----
+
In this example, the release version is `106.2.1+up1.8.2`.

. Clone the `rancher/charts` repository containing the specific version of the `longhorn-crd` chart that matches your current release (for example, `106.2.1+up1.8.2` for Longhorn v1.8.2):
+
[,shell]
----

> `+git clone https://github.com/rancher/charts.git+`

Reviewer 2: ```suggestion
git clone https://github.com/rancher/charts.git
```

=== Authentication

* To authenticate Helm with SUSE Application Collection OCI chart repository, refer to the https://docs.apps.rancher.io/get-started/authentication/#helm[Helm authentication documentation].
* To allow Kubernetes to pull images from the registry, follow the https://docs.apps.rancher.io/get-started/authentication/#kubernetes[Kubernetes authentication documentation] to create a secret and set the `global.imagePullSecrets` chart value.

== Migrating Longhorn Deployed via Helm To {longhorn-product-name}

If your existing Longhorn installation was deployed via Helm, you can verify it by running `helm list --all-namespaces`. By default, the Helm release used to install Longhorn is named `longhorn`, but it may also be a custom name defined by the user.
	
Next, optionally create a `values.yaml` file with any customized settings you want to apply during the upgrade.

Proceed with the upgrade to {longhorn-product-name} using the {longhorn-product-name} chart:

[,shell]
----
helm upgrade longhorn oci://dp.apps.rancher.io/charts/suse-storage \
	--namespace longhorn-system \

> 	--version <new-version> \

Reviewer 2: ```suggestion
	--version {patch-version} \
```

	--set global.imagePullSecrets=<PULL_IMAGE_SECRET>
----

Optionally, if you have a `values.yaml` file with customized settings, include it with the `-f` flag:

[,shell]
----
helm upgrade longhorn oci://dp.apps.rancher.io/charts/suse-storage \
	--namespace longhorn-system \

> 	--version <new-version> \

Reviewer 2: ```suggestion
	--version {patch-version} \
```

+
[,shell]
----
git clone https://github.com/rancher/charts.git
----
+
Navigate into the directory corresponding to your `longhorn-crd` chart version within the cloned repository.
+
Or, you can execute this link:{attachmentsdir}/download-longhorn-crd-chart.sh[script] to `longhorn-crd` chart.
+
After running the command `bash ./download-longhorn-crd-chart.sh 106.2.1+up1.8.2`, the `longhorn-crd` chart will be downloaded to the local directory `106.2.1+up1.8.2`. Then, open `106.2.1+up1.8.2/Chart.yaml` to verify that the chart version matches `106.2.1+up1.8.2`.

. Manually patch the `helm.sh/resource-policy: keep` annotation to *each Custom Resource Definitions (CRDs)* in `templates/crds.yaml` within the `longhorn-crd` chart that you cloned. This ensures that Helm does not delete the CRDs when the release is uninstalled.
Alternatively, you can use the link:{attachmentsdir}/patch-resource-policy-annotation.sh[provided script] to automatically patch the annotation:
+
[,shell]
----
bash ./patch-resource-policy-annotation.sh 106.2.1+up1.8.2/templates/crds.yaml
----
+
To verify that the CRDs have been correctly patched, run:
+
[,shell]
----

> vimdiff 106.2.1+up1.8.2/templates/crds.yaml.original 106.2.1+up1.8.2/templates/crds.yaml

Reviewer 2: ```suggestion
vimdiff /tmp/crds.yaml.original 106.2.1+up1.8.2/templates/crds.yaml
```

[CustomResourceDefinition] backingimages.longhorn.io
[CustomResourceDefinition] volumeattachments.longhorn.io
[CustomResourceDefinition] volumes.longhorn.io
[CustomResourceDefinition] backupbackingimages.longhorn.io
[CustomResourceDefinition] backups.longhorn.io
[CustomResourceDefinition] backuptargets.longhorn.io
[CustomResourceDefinition] backupvolumes.longhorn.io
[CustomResourceDefinition] engineimages.longhorn.io
[CustomResourceDefinition] engines.longhorn.io
[CustomResourceDefinition] instancemanagers.longhorn.io
----

=== Replace `longhorn-crd` with `longhorn` in Longhorn CRDs

After ensuring the `longhorn-crd` Helm release is uninstalled but the CRDs are kept, you need to update the ownership labels on the existing Longhorn CRDs to prepare for the main `longhorn` Helm chart. Apply the link:{attachmentsdir}/migrate-crd-ownership.sh[shell script] to perform the replacement.

=== Upgrade to {longhorn-product-name}

After the CRDs have been prepared, you can proceed with upgrading your Longhorn installation to {longhorn-product-name}:

[,shell]
----
helm upgrade longhorn oci://dp.apps.rancher.io/charts/suse-storage \
	--namespace longhorn-system \

> 	--version <new-version> \

Reviewer 2: ```suggestion
	--version {patch-version} \
```

	--set global.imagePullSecrets=<PULL_IMAGE_SECRET>
----

Optionally, if you have a `values.yaml` file with customized settings, include it with the `-f` flag:

[,shell]
----
helm upgrade longhorn oci://dp.apps.rancher.io/charts/suse-storage \
	--namespace longhorn-system \

> 	--version <new-version> \

Reviewer 2: ```suggestion
	--version {patch-version} \
```

[CustomResourceDefinition] settings.longhorn.io
[CustomResourceDefinition] sharemanagers.longhorn.io
[CustomResourceDefinition] snapshots.longhorn.io
[CustomResourceDefinition] supportbundles.longhorn.io
[CustomResourceDefinition] systembackups.longhorn.io
[CustomResourceDefinition] systemrestores.longhorn.io
[CustomResourceDefinition] backingimages.longhorn.io
[CustomResourceDefinition] volumeattachments.longhorn.io
[CustomResourceDefinition] volumes.longhorn.io
[CustomResourceDefinition] backupbackingimages.longhorn.io
[CustomResourceDefinition] backups.longhorn.io
[CustomResourceDefinition] backuptargets.longhorn.io
[CustomResourceDefinition] backupvolumes.longhorn.io
[CustomResourceDefinition] engineimages.longhorn.io
[CustomResourceDefinition] engines.longhorn.io
[CustomResourceDefinition] instancemanagers.longhorn.io
----

=== Replace `longhorn-crd` with `longhorn` in Longhorn CRDs

After ensuring the `longhorn-crd` Helm release is uninstalled but the CRDs are kept, you need to update the ownership labels on the existing Longhorn CRDs to prepare for the main `longhorn` Helm chart. Apply the following script to perform this replacement:

[,bash]
----
#!/bin/bash

> # List of CRDs to update
> Suggested change

Reviewer: ```suggestion
# The list of CRDs needs to be changed to reflect your installed Longhorn CRDs.
```

      "volumes.longhorn.io")

# Function to update a single CRD
update_crd() {
  local crd=$1
  echo "Processing CRD: $crd"

  # Get the current CRD definition
  kubectl get crd "$crd" -o yaml > temp-crd.yaml

  # Check if the CRD exists
  if [ $? -ne 0 ]; then
    echo "Error: CRD $crd not found"
    rm -f temp-crd.yaml
    return 1
  fi

  # Use sed with cross-platform compatible syntax
  # Create a backup file and replace longhorn-crd with longhorn
  sed -e 's/longhorn-crd/longhorn/g' temp-crd.yaml > temp-crd-updated.yaml

  if [ $? -ne 0 ]; then
    echo "Error: sed command failed for $crd"
    rm -f temp-crd.yaml temp-crd-updated.yaml
    return 1
  fi

  # Move updated file back to original
  mv temp-crd-updated.yaml temp-crd.yaml

  # Apply the updated CRD
  kubectl apply -f temp-crd.yaml

  if [ $? -eq 0 ]; then
    echo "Successfully updated CRD: $crd"
  else
    echo "Error updating CRD: $crd"
  fi

  # Clean up
  rm -f temp-crd.yaml
}

# Main execution
for crd in "${CRDS[@]}"; do
  update_crd "$crd"
done
----

=== Upgrade to {longhorn-product-name}

After the CRDs have been prepared, you can proceed with upgrading your Longhorn installation to {longhorn-product-name}.

> . Clone the {longhorn-product-name} Longhorn chart, or optionally prepare your `values.yaml` file from the `longhorn` chart repository. This step ensures you have the correct chart files locally for the upgrade.

Reviewer: Assume, should be able to upgrade from the OCI registry directly.

= Migrating to {longhorn-product-name}
	
== Migrating Longhorn Deployed via Helm To {longhorn-product-name}
	
If your existing Longhorn installation was deployed via Helm, you can verify it by running `helm list --all-namespaces`, where you should see a single Helm release named `longhorn`.
	
Next, optionally prepare a `values.yaml` file from your current Longhorn chart for the upgrade.

Proceed to upgrade to {longhorn-product-name} using the Helm chart you have prepared:

[,shell]
----
helm upgrade longhorn -n longhorn-system ./
----
	
== Migrating Longhorn Deployed via Rancher Apps Marketplace To {longhorn-product-name}

If your existing Longhorn installation was deployed via Rancher Apps Marketplace, you will see two Helm releases associated with Longhorn when running `helm list --all-namespaces | grep longhorn`.

> [,shell]
> ----
> $ helm list --all-namespaces | grep longhorn
> longhorn                                         	longhorn-system    	1       	2025-06-26 08:04:10.877379704 +0000 UTC	deployed	longhorn-105.1.0+up1.7.2                                                                               	v1.7.2
> longhorn-crd                                     	longhorn-system    	1       	2025-06-26 08:04:08.322051059 +0000 UTC	deployed	longhorn-crd-105.1.0+up1.7.2                                                                           	v1.7.2
> ----

Reviewer 2: ```suggestion
[,shell]
----
$ helm list --all-namespaces | grep longhorn
longhorn                                         	longhorn-system    	1       	2025-07-21 02:58:40.877484249 +0000 UTC	deployed	longhorn-106.2.1+up1.8.2                                                                               	v1.8.2
longhorn-crd                                     	longhorn-system    	1       	2025-07-21 02:58:38.192579711 +0000 UTC	deployed	longhorn-crd-106.2.1+up1.8.2                                                                           	v1.8.2
----
```

To migrate the existing Rancher Longhorn installation to {longhorn-product-name}, you need to delete the `longhorn-crd` Helm release without removing the CRDs, and then upgrade the `longhorn` Helm release by following the steps below. 

=== Remove `longhorn-crd` Helm Release while Keeping CRDs

To remove the `longhorn-crd` Helm release from your system while ensuring the Longhorn CRDs themselves are preserved, follow these steps:

. Check the current version of your `longhorn-crd` Helm release:
+
[,shell]
----
helm list --all-namespaces | grep longhorn-crd
----
+
The output will show your `longhorn-crd` release details, for example:
+

> [,bash]
> ----
> longhorn-crd                                     	longhorn-system    	1       	2025-06-26 08:04:08.322051059 +0000 UTC	deployed	longhorn-crd-105.1.0+up1.7.2
> ----

Reviewer 2: ```suggestion
[,bash]
----
longhorn-crd                                     	longhorn-system    	1       	2025-07-21 02:58:38.192579711 +0000 UTC	deployed	longhorn-crd-106.2.1+up1.8.2                                                                           	v1.8.2
----
```

+

> In this example, the release version is `105.1.0+up1.7.2`.

Reviewer 2: ```suggestion
In this example, the release version is `106.2.1+up1.8.2`.
```

> . Clone the `rancher/charts` repository containing the specific version of the `longhorn-crd` chart that matches your current release (for example, `105.1.0+up1.7.2` for Longhorn v1.7.2):

Reviewer 2: ```suggestion
. Clone the `rancher/charts` repository containing the specific version of the `longhorn-crd` chart that matches your current release (for example, `106.2.1+up1.8.2` for Longhorn v1.8.2):
```

+
`+git clone https://github.com/rancher/charts.git+`
+
Navigate into the directory corresponding to your `longhorn-crd` chart version within the cloned repository.
+
You can execute the script to download the `longhorn-crd` chart.
+
[,shell]
----
#!/bin/bash
set -euo pipefail
if [ $# -lt 1 ] || [ $# -gt 2 ]; then
    echo "Usage: $0 <version> [github_token]"
    exit 1
fi
REPO="rancher/charts"
VERSION="$1"
TARGET_PATH="charts/longhorn-crd/${VERSION}"

> LOCAL_DIR="chart"

Reviewer 2: ```suggestion
LOCAL_DIR=""$VERSION""
```

        elif [[ "${type}" == "dir" ]]; then
            echo "Entering directory: ${path}"
            mkdir -p "${local_path}/${name}"
            download_directory "${path}" "${local_path}/${name}" "${branch}"
        fi
    done
}
for branch in $branches; do
    echo "Checking branch: $branch"
    url="https://api.github.com/repos/${REPO}/contents/${TARGET_PATH}?ref=${branch}"
    if [ -n "$AUTH_HEADER" ]; then
        response=$(curl -s -w "%{http_code}" -H "${AUTH_HEADER}" -o /tmp/check_response.json "$url")
    else
        response=$(curl -s -w "%{http_code}" -o /tmp/check_response.json "$url")
    fi
    http_code="${response: -3}"
    body=$(cat /tmp/check_response.json)
    if [[ "$http_code" == "200" ]]; then
        echo "Found target in branch: $branch"
        mkdir -p "${LOCAL_DIR}"
        download_directory "${TARGET_PATH}" "${LOCAL_DIR}" "${branch}"
        echo "Download completed to ${LOCAL_DIR}."
        exit 0
    elif [[ "$http_code" != "404" ]]; then
        echo "Error checking branch: ${branch}"
        echo "HTTP status: $http_code"
        echo "Response: $body"
        exit 1
    fi
done
echo "Target not found in any release-* branch."
exit 1
----

> After running the command `bash ./download_longhorn-crd_chart.sh 105.1.0+up1.7.2`, the `longhorn-crd` chart will be downloaded to the local directory `105.1.0+up1.7.2`.

Reviewer 2: ```suggestion
After running the command `bash ./download_longhorn-crd_chart.sh 106.2.1+up1.8.2`, the `longhorn-crd` chart will be downloaded to the local directory `106.2.1+up1.8.2`. Then, open `106.2.1+up1.8.2/Chart.yaml` to verify that the chart version matches `106.2.1+up1.8.2`.
```

= Migrating to {longhorn-product-name}
	
== Migrating Longhorn Deployed via Helm To {longhorn-product-name}
	

> [NOTE]

Reviewer 2: Can we use `Notice`?
